### PR TITLE
Morphology

### PR DESCRIPTION
**Pull request by Juclique**
### Description
Added the basic functions for morphology module: dilation and erosion

### Status
**Ready**

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/morphology/black_hat.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+           tensor (torch.Tensor): Image with shape :math:`(B, C, H, W)`.
+           kernel (torch.Tensor): Structuring element with shape :math:`(H, W)`.
+
+        Returns:
+           torch.Tensor: Top hat transformated image with shape :math:`(B, C, H, W)`.
+
+        Example:
+            >>> tensor = torch.rand(1, 3, 5, 5)
+            >>> kernel = torch.ones(3, 3)
+            >>> black_hat_img = black_hat(tensor, kernel)
```
Comment (edgarriba) [inline kornia/morphology/black_hat.py:24]: `kornia.morphology.black_hat_img = black_hat(tensor, kernel)` we'll add soon doctests that probably will make this to crash

File: kornia/morphology/black_hat.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+from kornia.morphology import close
```
Comment (edgarriba) [inline kornia/morphology/black_hat.py:1]: consider importing module: https://softwareengineering.stackexchange.com/questions/187403/import-module-vs-from-module-import-function

File: kornia/morphology/dilation.py [hunk 3]
```diff
@@ -24,14 +24,14 @@ def se_to_mask(se):
 
         self.kernel = se_to_mask(self.se)
 
-    def forward(self, input):
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/morphology/dilation.py:27]: consider adding assertion with meaningful error messages

File: kornia/morphology/gradient.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        Returns:
+           torch.Tensor: Dilated image with shape :math:`(B, C, H, W)`.
+
+        Example:
+            >>> tensor = torch.rand(1, 3, 5, 5)
+            >>> kernel = torch.ones(3, 3)
+            >>> gradient_img = gradient(tensor, kernel)
+        """
+
+    return Dilate(kernel)(tensor) - Erode(kernel)(tensor)
```
Comment (edgarriba) [inline kornia/morphology/gradient.py:24]: isn't this similar to spatial_gradient ? /cc @ducha-aiki 
https://kornia.readthedocs.io/en/latest/filters.html?highlight=spatial%20gradient#kornia.filters.spatial_gradient

File: test/morphology/test_black_hat.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import pytest
+import torch
+import kornia.morphology.black_hat as black_hat
+import kornia.testing as utils  # test utils
+from torch.autograd import gradcheck
+
+
+class TestBlackHat:
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:8]: inherit from `BaseTest`. See: https://github.com/kornia/kornia/blob/master/test/enhance/test_normalize.py#L223

File: test/morphology/test_black_hat.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    def test_shape_channels(self, device):
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:9]: you have to pass `dtype`

File: test/morphology/test_black_hat.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        input = torch.rand(1, 3, 4, 6).to(device)
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:10]: tensor must be initialized as `torch.tensor(..., device=device, dtype=dtype)`

File: test/morphology/test_black_hat.py (hunk 5 above)
Comment (edgarriba) [inline test/morphology/test_black_hat.py:8]: I miss some numeric unit tests. maybe you can generate some values from opencv

**Review by edgarriba — CHANGES_REQUESTED** [review]
Good overall. Checkout my comments. Don't forget to add to the docs

File: kornia/morphology/dilation.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    Returns:
+       torch.Tensor: Dilated image with shape :math:`(B, C, H, W)`.
+
+    Example:
+        >>> tensor = torch.rand(1, 3, 5, 5)
+        >>> kernel = torch.ones(3, 3)
+        >>> dilated_img = dilation(tensor, kernel)
+    """
+    return Dilate(kernel)(tensor)
```
Comment (ducha-aiki) [inline kornia/morphology/dilation.py:52]: I know that we are not consistent in kornia ourselves, but it is better to have module, which calls function, rather than function, which creates module every time. 

E.g. see here: 
https://github.com/kornia/kornia/blob/master/kornia/feature/responses.py#L347

File: kornia/morphology/dilation.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+class Dilate(nn.Module):
+
+    def __init__(self, se: torch.Tensor) -> None:
+        super().__init__()
+        self.se = se - 1
+        self.se_h, self.se_w = se.shape
+        self.pad = (self.se_h // 2, self.se_w // 2)
+
+        def se_to_mask(se: torch.Tensor) -> torch.Tensor:
```
Comment (ducha-aiki) [inline kornia/morphology/dilation.py:14]: Better put se_to_mask as separate util function, which is called by both erode and dillate. Otherwise it is code duplication

File: kornia/morphology/gradient.py (hunk 4 above)
Comment (ducha-aiki) [reply kornia/morphology/gradient.py:24]: @edgarriba  it is similar, but not the same and  it is better fit morphology module. So let it be here.

Comment (ducha-aiki) [inline kornia/morphology/gradient.py:24]: See comment about module calling function

File: kornia/morphology/top_hat.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
+from kornia.morphology import open
+import torch
+
+
+def top_hat(tensor: torch.Tensor, kernel: torch.Tensor) -> torch.Tensor:
```
Comment (ducha-aiki) [inline kornia/morphology/top_hat.py:5]: Logistic question: why do you prefer creating a separate file for each operation?

File: test/morphology/test_black_hat.py (hunk 5 above)
Comment (ducha-aiki) [reply test/morphology/test_black_hat.py:8]: +1 for adding numerical tests.

**Review by ducha-aiki — CHANGES_REQUESTED** [review]

File: test/morphology/test_black_hat.py (hunk 5 above)
Comment (Juclique) [reply test/morphology/test_black_hat.py]: Done, but I turned off JIT.

Comment (Juclique) [reply test/morphology/test_black_hat.py]: I created a small test, but with custom numbers. That could be enough?

File: kornia/morphology/top_hat.py (hunk 10 above)
Comment (Juclique) [reply kornia/morphology/top_hat.py]: No particular reason, i just created one after another.
I ordered the scripts in similar functions (dilate/erode) (open/close) (others). 
Could be better in that way?

File: .idea/.gitignore [hunk 11]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (edgarriba) [inline .idea/.gitignore:1]: remove this file from the PR

File: .idea/Kornia.iml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edgarriba) [inline .idea/Kornia.iml:1]: remove this file from the PR

File: .idea/inspectionProfiles/profiles_settings.xml [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
```
Comment (edgarriba) [inline .idea/inspectionProfiles/profiles_settings.xml:1]: remove this file from the PR

File: .idea/misc.xml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edgarriba) [inline .idea/misc.xml:1]: remove this file from the PR

File: .idea/modules.xml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edgarriba) [inline .idea/modules.xml:1]: remove this file from the PR

File: .idea/vcs.xml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edgarriba) [inline .idea/vcs.xml:1]: remove this file from the PR

File: docs/.gitignore [hunk 17]
```diff
@@ -1,5 +1,3 @@
-build/doctest
```
Comment (edgarriba) [inline docs/.gitignore:1]: keep it

File: kornia/morphology/basic_operators.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
+# Libraries
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+# se_to_mask
+def se_to_mask(se: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/morphology/basic_operators.py:8]: I suggest naming this functiona as `_se_to_mask`since it's not part of the main api

File: kornia/morphology/basic_operators.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+       tensor (torch.Tensor): Image with shape :math:`(B, C, H, W)`.
+       kernel (torch.Tensor): Structuring element with shape :math:`(H, W)`.
+
+    Returns:
+       torch.Tensor: Dilated image with shape :math:`(B, C, H, W)`.
+
+    Example:
+        >>> tensor = torch.rand(1, 3, 5, 5)
+        >>> kernel = torch.ones(3, 3)
+        >>> dilated_img = kornia.morphology.dilation(tensor, kernel)
```
Comment (edgarriba) [inline kornia/morphology/basic_operators.py:53]: I know I explicitly asked to change this, but we managed to integrate doctest and seems that the way it works, the functionalities cannot have the full namespace. Rename it to: `>>> dilated_img = dilation(tensor, kernel)`

To test this, first you need to add the morphology module to here:
https://github.com/kornia/kornia/blob/master/Makefile#L42 and then call `make doctest`.
Sorry for the inconvenience, we are just starting to make robust this feature and discovering its usage on the fly.

File: kornia/morphology/basic_operators.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        output = input.view(input.shape[0] * input.shape[1], 1, input.shape[2], input.shape[3])
+        output = (F.conv2d(output, self.kernel, padding=self.pad) + self.se.view(1, -1, 1, 1)).max(dim=1)[0]
+
+        return output.view(*input.shape)
+
+
+def dilation(tensor: torch.Tensor, kernel: torch.Tensor) -> torch.Tensor:
+
+    r"""
+    Returns the dilated image applying the same kernel in each channel.
```
Comment (edgarriba) [inline kornia/morphology/basic_operators.py:40]: add an end line between the main functon description and details. Same for other functions

File: kornia/morphology/basic_operators.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    The kernel must have 2 dimensions, each one defined by an odd number.
+
+    Args
+       tensor (torch.Tensor): Image with shape :math:`(B, C, H, W)`.
+       kernel (torch.Tensor): Structuring element with shape :math:`(H, W)`.
```
Comment (edgarriba) [inline kornia/morphology/basic_operators.py:45]: question: can we support (B, H, W) kernel? Now I guess it's a single kernel for the whole batch. /cc @ducha-aiki  

BTW, HW might be confused with the HW of the image. Use kx, ky or similar. Check other similar functionalities and make it compliant to it.

File: kornia/morphology/basic_operators.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+
+    Returns:
+       torch.Tensor: Eroded image with shape :math:`(B, C, H, W)`.
+
+    Example:
+        >>> tensor = torch.rand(1, 3, 5, 5)
+        >>> kernel = torch.ones(5, 5)
+        >>> output = kornia.morphology.erosion(tensor, kernel)
```
Comment (edgarriba) [inline kornia/morphology/basic_operators.py:107]: same comments as before for docs in general

File: kornia/morphology/morphology.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
+# Libraries
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import kornia.morphology as m
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:5]: to avoid confusions and potential issues with debuggers, use a better module name e.g `morph`

File: test/morphology/test_black_hat.py [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+        with pytest.raises(ValueError):
+            test = torch.ones(2, 3, 4)
+            assert m.black_hat(input, test)
+
+    def test_gradcheck(self, device, dtype):
+        input = torch.rand(2, 3, 4, 4, requires_grad=True, device=device, dtype=dtype).double()
+        kernel = torch.rand(3, 3, requires_grad=True, device=device, dtype=dtype).double()
+        assert gradcheck(m.black_hat, (input, kernel), raise_exception=True)
+
+    @pytest.mark.skip(reason="turn off all jit for a while")
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:52]: turn on

File: test/morphology/test_black_hat.py [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+        op = m.black_hat
+        op_script = torch.jit.script(op)
+
+        input = torch.rand(1, 2, 7, 7, device=device, dtype=dtype)
+        kernel = torch.ones(3, 3, device=device, dtype=dtype)
+
+        actual = op_script(input, kernel)
+        expected = op(input, kernel)
+
+        assert_allclose(actual, expected)
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:63]: add `test_module` to test that the modules and functionals behave in the same way

Comment (edgarriba) [reply test/morphology/test_black_hat.py:63]: check last implementation for kornia color test which has new name sintaxis too

File: test/morphology/test_black_hat.py [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+
+class TestBlackHat(utils.BaseTester):
+
+    def test_smoke(self, device, dtype):
+        kernel = torch.rand(3, 3, device=device, dtype=dtype)
+        assert m.se_to_mask(kernel) is not None
+
+    def test_batch(self, device, dtype):
+        input = torch.rand(3, 2, 6, 10, device=device, dtype=dtype)
+        kernel = torch.rand(3, 3, device=device, dtype=dtype)
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:17]: I suggest to add parametrize with different image and kernels size (check for even shapes as common corner cases). Check kornia color test_cardinality

File: test/morphology/test_black_hat.py [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+        test = m.black_hat(input, kernel)
+        assert input.shape == test.shape == (3, 2, 6, 10)
+
+    def test_value(self, device, dtype):
+        input = torch.tensor([[0.5, 1., 0.3], [0.7, 0.3, 0.8], [0.4, 0.9, 0.2]],
+                             device=device, dtype=dtype).unsqueeze(0).unsqueeze(0)
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:23]: can you try better formatting ? to make more readable the tests? use `[]` instead of unsqueeze, it's more pythonic

File: test/morphology/test_black_hat.py [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+        kernel = torch.tensor([[0., 1., 0.], [1., 1., 1.], [0., 1., 0.]], device=device, dtype=dtype)
+        expected = torch.tensor([[0.2, 0., 0.5], [0., 0.4, 0.], [0.3, 0., 0.6]],
+                                device=device, dtype=dtype).unsqueeze(0).unsqueeze(0)
+        assert_allclose(m.black_hat(input, kernel), expected)
+
+    def test_exception(self, device, dtype):
+        input = torch.ones(1, 1, 3, 4, device=device, dtype=dtype).double()
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:30]: why double dtype ?

File: test/morphology/test_black_hat.py [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+        kernel = torch.ones(3, 3, device=device, dtype=dtype).double()
+
+        with pytest.raises(TypeError):
+            assert m.black_hat([0.], kernel)
+
+        with pytest.raises(TypeError):
+            assert m.black_hat(input, [0.])
+
+        with pytest.raises(ValueError):
+            test = torch.ones(2, 3, 4)
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:40]: device dtype

File: test/morphology/test_black_hat.py [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+            assert m.black_hat(test, kernel)
+
+        with pytest.raises(ValueError):
+            test = torch.ones(2, 3, 4)
+            assert m.black_hat(input, test)
+
+    def test_gradcheck(self, device, dtype):
+        input = torch.rand(2, 3, 4, 4, requires_grad=True, device=device, dtype=dtype).double()
```
Comment (edgarriba) [inline test/morphology/test_black_hat.py:48]: dtype=torch.float64 and remove double

**Review by edgarriba — CHANGES_REQUESTED** [review]

**Comment by edgarriba** [comment]
@Juclique We are releasing v0.5 in a week - would be great to include this feature.

**Comment by edgarriba** [comment]
@Juclique what happened here ? all the code is gone
